python/singa/sonnx.py

```diff
@@ -0,0 +1,289 @@
+        handledic[name].device_id = x.device.id()
+        return handledic
+
+
+
+
+
+
+    @staticmethod
+    def onnx_model_init(model,device):
+        '''
+        input model
+
+        return: model and model dictionary
+        '''
+
+        modeldic = {}
+        for i in model.graph.node:
+            if (i.op_type == 'Constant'):
+                modeldic[str(i.output[0])] = tensor.Tensor(device=device,data=numpy_helper.to_array(i.attribute[0].t),requires_grad=True, stores_grad=True)
+
+        return model,modeldic
+
+
+    @staticmethod
+    def find_name(model,name):
+        for i in model.graph.node:
+            if (i.name == name):
+                return i
+
+
+    @staticmethod
+    def find_shape(model,input):
+        '''
+        # find weight shape for layers
+        '''
+        for i in model.graph.node:
+            if (i.op_type == 'Constant' and i.output[0] == input):
+                return numpy_helper.to_array(i.attribute[0].t).shape
+
+
+    @staticmethod
+    def run_model(model,inputs,device):
+        model, modeldic  = Backend.onnx_model_init(model,device)
+        return Backend.run(model, modeldic,inputs)[0]
+
+    @staticmethod
+    def run(model, modeldic,inputs,handledic={}):
+        '''
+            input: input for singa model
+            load other nodes of onnx
+            '''
+        supportLayer = ['Conv','MaxPool','AveragePool','BatchNormalization']
+        oper=modeldic
+        autograd.training = True
+        for counter,i in enumerate(model.graph.input):
+            oper[i.name] = inputs[counter]
+        for i in model.graph.node:
+            if (i.op_type == 'Relu'):
+                oper[str(i.output[0])] = autograd.relu(oper[str(i.input[0])])
+            elif (i.op_type == 'Softmax'):
+                oper[str(i.output[0])] = autograd.softmax(oper[str(i.input[0])])
+            elif (i.op_type == 'Add'):
+                oper[str(i.output[0])] = autograd.add(oper[str(i.input[0])], oper[str(i.input[1])])
+            elif (i.op_type == 'MatMul'):
+                oper[str(i.output[0])] = autograd.matmul(oper[str(i.input[0])], oper[str(i.input[1])])
+            elif (i.op_type == 'Flatten'):
+                oper[str(i.output[0])] = autograd.flatten(oper[str(i.input[0])])
+            elif(i.op_type == 'Concat'):
+                oper[str(i.output[0])] = autograd.cat((oper[str(i.input[0])], oper[str(i.input[1])]),int(i.attribute[0].i))
+            elif(i.op_type == 'Tanh'):
+                oper[str(i.output[0])] = autograd.tanh(oper[str(i.input[0])])
+            elif (i.op_type == 'Sigmoid'):
+                oper[str(i.output[0])] = autograd.sigmoid(oper[str(i.input[0])])
+            elif (i.op_type == 'Mul'):
+                oper[str(i.output[0])] = autograd.mul(oper[str(i.input[0])],oper[str(i.input[1])])
+            elif (i.op_type == 'Conv'):
+                handledic = Backend.convhandle(i.name,handledic,oper[str(i.input[0])],model)
```

Comment:
if the handle is used immediately, do you need to put it into a dict?
creating the handle in every iteration is expensive!

Reply:
I had written a if check :   
` if(name in handledic):return handledic`
if I create handle before, will not create handle again